Support 5.8 kernels as on Fedora 32

elastio-snap build is failing on Fedora 32 with the kernel 5.8.9-200.
```
make -C /lib/modules/5.8.9-200.fc32.x86_64/build M=/home/elastio/elastio-snap/src modules
make[2]: Entering directory '/usr/src/kernels/5.8.9-200.fc32.x86_64'
  CC [M]  /home/elastio/elastio-snap/src/elastio-snap.o
In file included from ./include/linux/umh.h:4,
                 from ./include/linux/kmod.h:9,
                 from ./include/linux/module.h:16,
                 from /home/elastio/elastio-snap/src/includes.h:11,
                 from /home/elastio/elastio-snap/src/elastio-snap.c:8:
/home/elastio/elastio-snap/src/elastio-snap.c: In function ‘__tracer_setup_snap’:
./include/linux/gfp.h:297:20: warning: passing argument 1 of ‘blk_alloc_queue’ makes pointer from integer without a cast [-Wint-conversion]
  297 | #define GFP_KERNEL (__GFP_RECLAIM | __GFP_IO | __GFP_FS)
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                    |
      |                    unsigned int
/home/elastio/elastio-snap/src/elastio-snap.c:3584:34: note: in expansion of macro ‘GFP_KERNEL’
 3584 |  dev->sd_queue = blk_alloc_queue(GFP_KERNEL);
      |                                  ^~~~~~~~~~
In file included from /home/elastio/elastio-snap/src/includes.h:13,
                 from /home/elastio/elastio-snap/src/elastio-snap.c:8:
./include/linux/blkdev.h:1172:55: note: expected ‘blk_qc_t (*)(struct request_queue *, struct bio *)’ {aka ‘unsigned int (*)(struct request_queue *, struct bio *)’} but argument is of type ‘unsigned int’
 1172 | struct request_queue *blk_alloc_queue(make_request_fn make_request, int node_id);
      |                                       ~~~~~~~~~~~~~~~~^~~~~~~~~~~~
/home/elastio/elastio-snap/src/elastio-snap.c:3584:18: error: too few arguments to function ‘blk_alloc_queue’
 3584 |  dev->sd_queue = blk_alloc_queue(GFP_KERNEL);
      |                  ^~~~~~~~~~~~~~~
In file included from /home/elastio/elastio-snap/src/includes.h:13,
                 from /home/elastio/elastio-snap/src/elastio-snap.c:8:
./include/linux/blkdev.h:1172:23: note: declared here
 1172 | struct request_queue *blk_alloc_queue(make_request_fn make_request, int node_id);
      |                       ^~~~~~~~~~~~~~~
/home/elastio/elastio-snap/src/elastio-snap.c:3593:2: error: implicit declaration of function ‘blk_queue_make_request’; did you mean ‘blk_queue_max_segments’? [-Werror=implicit-function-declaration]
 3593 |  blk_queue_make_request(dev->sd_queue, snap_mrf);
      |  ^~~~~~~~~~~~~~~~~~~~~~
      |  blk_queue_max_segments
cc1: some warnings being treated as errors
make[3]: *** [scripts/Makefile.build:281: /home/elastio/elastio-snap/src/elastio-snap.o] Error 1
make[2]: *** [Makefile:1756: /home/elastio/elastio-snap/src] Error 2
make[2]: Leaving directory '/usr/src/kernels/5.8.9-200.fc32.x86_64'
make[1]: *** [Makefile:14: default] Error 2
make[1]: Leaving directory '/home/elastio/elastio-snap/src'
make: *** [Makefile:24: driver] Error 2
```

--- a/.github/scripts/set_env.sh
+++ b/.github/scripts/set_env.sh
@@ -19,3 +19,4 @@ echo "RUNNER_NUM=$runner_num"                           >> $GITHUB_ENV
 echo "BOX_DIR=.github/buildbox"                         >> $GITHUB_ENV
 echo "BOX_NAME=$DISTRO-$ARCH-build"                     >> $GITHUB_ENV
 echo "INSTANCE_NAME=$DISTRO-$ARCH-build-$runner_num"    >> $GITHUB_ENV
+echo "TEST_IMAGE=$PWD/$DISTRO-test_image.qcow2"         >> $GITHUB_ENV

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -1,6 +1,6 @@
 ---
 name: ci
-on: 
+on:
   - push
 
 env:
@@ -20,13 +20,16 @@ jobs:
           debian8, debian9, debian10,
           centos7, centos8,
           amazon2,
-          # TODO: Bring back Fedora when issue https://github.com/elastio/elastio-snap/issues/44 is fixed
-          # fedora31, fedora32
+          fedora31
+          # TODO: Bring back
+          # - Fedora 32 when kernel 5.9  is supported (issue https://github.com/elastio/elastio-snap/issues/57 is fixed)
+          # - Fedora 33 when kernel 5.10 is supported.
+          # fedora32, fedora33
         ]
         arch: [ amd64 ]
 
     steps:
-      - name: Checkout sources 
+      - name: Checkout sources
         uses: actions/checkout@v2
 
       - name: Set ENV
@@ -62,12 +65,31 @@ jobs:
         run: vagrant ssh ${{env.INSTANCE_NAME}} -c 'sudo make install'
         working-directory: ${{env.BOX_DIR}}
 
+        # NOTE: This special logic with the real block device instead of the loop device is used especially for Fedora 31
+        # due to ussie https://github.com/elastio/elastio-snap/issues/71. It can be simplified back after the issue is fixed.
+      - name: Attach external drive
+        if: ${{ matrix.distro == 'fedora31' }}
+        run: |
+          qemu-img create -f qcow2 ${TEST_IMAGE} 1G
+          virsh attach-disk --domain ${BOX_DIR##*/}_${INSTANCE_NAME} --source ${TEST_IMAGE} --target vdb --driver qemu --subdriver qcow2 --targetbus virtio
+          cd $BOX_DIR && vagrant ssh ${{env.INSTANCE_NAME}} -c 'echo -e "n\np\n\n\n\nw" | sudo fdisk /dev/vdb'
+
       - name: Run tests
-        run: vagrant ssh ${{env.INSTANCE_NAME}} -c 'cd tests && sudo ./elio-test.sh'
+        run: |
+          [[ $INSTANCE_NAME == fedora31* ]] && device=/dev/vdb1
+          vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh $device"
         working-directory: ${{env.BOX_DIR}}
         # For now tests are taking 10-20 seconds. But they can hang.
         # 5 minutes seems to be reasonable timeout.
         timeout-minutes: 5
+
+      - name: Detach external drive
+        if: always()
+        run: |
+          if virsh domblklist ${BOX_DIR##*/}_${INSTANCE_NAME} --details | grep "file" | awk '{ print $NF }' | grep ${TEST_IMAGE} ; then
+              virsh detach-disk --domain ${BOX_DIR##*/}_${INSTANCE_NAME} ${TEST_IMAGE}
+          fi
+          rm -f ${TEST_IMAGE}
 
       - name: Upload artifacts
         run: vagrant ssh ${{env.INSTANCE_NAME}} -c 'AWS_SECRET_ACCESS_KEY='"'$AWS_SECRET_ACCESS_KEY'"' repobuild/upload.sh
@@ -88,7 +110,7 @@ jobs:
       - baremetal
 
     steps:
-      - name: Checkout sources 
+      - name: Checkout sources
         uses: actions/checkout@v2
 
       - name: Make manifest
@@ -102,12 +124,12 @@ jobs:
 
   dispatch-packaging-repo:
     name: Trigger repo upload
-    needs: manifest 
+    needs: manifest
     runs-on:
       - baremetal
 
     steps:
-      - name: Checkout sources 
+      - name: Checkout sources
         uses: actions/checkout@v2
 
       - name: Dispatch packaging repo

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -66,7 +66,7 @@ jobs:
         working-directory: ${{env.BOX_DIR}}
 
         # NOTE: This special logic with the real block device instead of the loop device is used especially for Fedora 31
-        # due to ussie https://github.com/elastio/elastio-snap/issues/71. It can be simplified back after the issue is fixed.
+        # due to issue https://github.com/elastio/elastio-snap/issues/71. It can be simplified back after the issue is fixed.
       - name: Attach external drive
         if: ${{ matrix.distro == 'fedora31' }}
         run: |

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -37,6 +37,10 @@ MODULE_VERSION(ELASTIO_SNAP_VERSION);
 #include <uapi/linux/mount.h>
 #endif
 
+#ifdef HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID
+#include "linux/blk-mq.h"
+#endif
+
 #ifndef HAVE_BIO_LIST
 //#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,30)
 struct bio_list {
@@ -544,6 +548,13 @@ static inline int elastio_snap_call_mrf(make_request_fn *fn, struct request_queu
 	#define MRF_RETURN_TYPE blk_qc_t
 	#define MRF_RETURN(ret) return BLK_QC_T_NONE
 
+#ifdef HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID
+static MRF_RETURN_TYPE elastio_snap_null_mrf(struct request_queue *q, struct bio *bio){
+	smp_wmb();
+	return blk_mq_make_request(q, bio);
+}
+#endif
+
 static inline int elastio_snap_call_mrf(make_request_fn *fn, struct request_queue *q, struct bio *bio){
 	return fn(q, bio);
 }
@@ -749,7 +760,7 @@ static inline void elastio_snap_bio_copy_dev(struct bio *dst, struct bio *src){
  *
  * Note: CentOS 7 has enum req_op starting from the version 7.4, kernel 3.10.0-693. But this enum has just 4 values
  * instead of 6 as in other kernels, where this enum is present. And it doesn't have defined REQ_OP_BITS, which could
- * could be defined and equal to the 2 bits.
+ * be defined and equal to the 2 bits.
  */
 #define __ELASTIO_SNAP_PASSTHROUGH 28	// set as the last flag bit
 #else
@@ -3127,9 +3138,12 @@ static MRF_RETURN_TYPE tracing_mrf(struct request_queue *q, struct bio *bio){
 	}
 
 call_orig:
+#ifndef HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID
 	if(orig_mrf) ret = elastio_snap_call_mrf(orig_mrf, q, bio);
 	else LOG_ERROR(-EFAULT, "error finding original_mrf");
-
+#else
+	ret = elastio_snap_call_mrf(orig_mrf, q, bio);
+#endif
 out:
 	MRF_RETURN(ret);
 }
@@ -3197,7 +3211,15 @@ static int find_orig_mrf(struct block_device *bdev, make_request_fn **mrf){
 	struct request_queue *q = bdev_get_queue(bdev);
 
 	if(q->make_request_fn != tracing_mrf){
+#ifndef HAVE_BLK_ALLOC_QUEUE_MK_REQ_FN_NODE_ID
 		*mrf = q->make_request_fn;
+#else
+		if (q->make_request_fn) *mrf = q->make_request_fn;
+		else{
+			*mrf = elastio_snap_null_mrf;
+			LOG_DEBUG("original mrf is empty, set to elastio_snap_null_mrf");
+		}
+#endif
 		return 0;
 	}
 
@@ -3427,7 +3449,7 @@ static int __tracer_destroy_cow(struct snap_device *dev, int close_method){
 	dev->sd_cache_size = 0;
 
 	if(dev->sd_cow){
-		LOG_DEBUG("destroying cow manager");
+		LOG_DEBUG("destroying cow manager. close method: %d", close_method);
 
 		if(close_method == 0){
 			cow_free(dev->sd_cow);
@@ -3630,7 +3652,7 @@ static int __tracer_setup_snap(struct snap_device *dev, unsigned int minor, stru
 		goto error;
 	}
 
-#ifdef HAVE_BLK_ALLOC_QUEUE_GFP_T 
+#ifdef HAVE_BLK_ALLOC_QUEUE_GFP_T
 	//register request handler
 	LOG_DEBUG("setting up make request function");
 	blk_queue_make_request(dev->sd_queue, snap_mrf);

--- a/tests/devicetestcase.py
+++ b/tests/devicetestcase.py
@@ -6,6 +6,7 @@
 #
 
 import os
+import subprocess
 import unittest
 
 import kmod
@@ -19,14 +20,18 @@ class DeviceTestCase(unittest.TestCase):
     def setUpClass(cls):
         cls.minor = randint(0, 23)
         r =  randint(0, 999)
-        cls.backing_store = "/tmp/disk_{0:03d}.img".format(r)
         cls.mount = "/tmp/elastio-snap_{0:03d}".format(r)
 
         cls.kmod = kmod.Module("../src/elastio-snap.ko")
         cls.kmod.load(debug=1)
-
-        util.dd("/dev/zero", cls.backing_store, 256, bs="1M")
-        cls.device = util.loop_create(cls.backing_store)
+        if os.getenv('TEST_DEVICE'):
+            cls.device = os.getenv('TEST_DEVICE')
+            dev_size = int(subprocess.check_output("blockdev --getsize64 %s" % cls.device, shell=True, text=True))//1024**2
+            util.dd("/dev/zero", cls.device, dev_size, bs="1M")
+        else:
+            cls.backing_store = "/tmp/disk_{0:03d}.img".format(r)
+            util.dd("/dev/zero", cls.backing_store, 256, bs="1M")
+            cls.device = util.loop_create(cls.backing_store)
 
         util.mkfs(cls.device)
         os.makedirs(cls.mount, exist_ok=True)
@@ -35,6 +40,7 @@ class DeviceTestCase(unittest.TestCase):
     @classmethod
     def tearDownClass(cls):
         util.unmount(cls.mount)
-        util.loop_destroy(cls.device)
-        os.unlink(cls.backing_store)
+        if hasattr(cls, "backing_store"):
+            util.loop_destroy(cls.device)
+            os.unlink(cls.backing_store)
         cls.kmod.unload()

--- a/tests/elio-test.sh
+++ b/tests/elio-test.sh
@@ -30,6 +30,13 @@ echo
 
 dmesg -c &> /dev/null
 >| dmesg.log
+
+[ ! -z "$1" ] && export TEST_DEVICE="$1"
+if ! lsblk $TEST_DEVICE >/dev/null 2>&1; then
+    echo "The script's argumet $TEST_DEVICE seems to be not a block device."
+    exit 1
+fi
+
 python3 -m unittest -v
 ret=$?
 dmesg > dmesg.log

